src/ast/analysis.rs

```diff
@@ -0,0 +1,100 @@
+//! Diverse checks that need to be done before lowering.
+use crate::{ast::{Located, Block, Node, Name, visitor::NodeVisitor}, ty::{Binding,Ty}};
+
+
+/// Checks if a function is recursive or not.
+pub struct RecursionChecker<'a> {
+    /// Name of the target function
+    name: Name<'a>,
+    /// Stores if the function is recursive or not in each step of the traversal.
+    is_rec: bool,
+    /// Stores if the target name is being shadowed in the current scope. It represents the top of
+    /// the stack
+    is_shadowed: bool,
+    /// Stores the shadowing status in the upper scopes.
+    stack: Vec<bool>,
+}
+
+impl<'a> RecursionChecker<'a> {
+    /// Runs the recursion check with the target function's name and body.
+    pub fn run(name: Name<'a>, body: &Block<'a>) -> bool {
+        let mut this = RecursionChecker {
+            name,
+            is_rec: false,
+            is_shadowed: false,
+            stack: Vec::new(),
+        };
+        this.visit_block(body);
+        // Sanity check. There should be only one scope after visiting the body function. the
+        // original one
+        assert!(this.stack.is_empty(), "Someone forgot to pop a scope from the stack");
+        this.is_rec
+    }
+
+    /// Push a new scope into the stack.
+    ///
+    /// The new scope has the same shadowed status as the old scope because names are preserved
+    /// when creating a new scope.
+    fn push_scope(&mut self) {
+        self.stack.push(self.is_shadowed)
+    }
+
+    /// Pops a scope from the stack.
+    ///
+    /// This function panics if there are no more scopes in the stack. Which should be impossible
+    /// because the stack always starts as non-empty and we should only pop newly added scopes from
+    /// the stack.
+    fn pop_scope(&mut self) {
+        self.is_shadowed = self.stack.pop().expect("there are no more scopes in the stack");
+    }
+}
+
+impl<'a> NodeVisitor<'a> for RecursionChecker<'a> {
+    fn visit_name(&mut self, name: &Name<'a>) {
+        // The function is recursive if its name is not shadowed in the current scope and we found
+        // it is somewhere inside its body.
+        if !self.is_shadowed && *name == self.name {
+            self.is_rec = true;
+        }
+        // Keep visiting
+        self.super_name(name);
```

Comment:
`super_name()` doesn't do anything so it is just there for completeness and symmetry with the other methods?

Reply:
Yes, I have to document the visitor trait to make that clear. The idea is that you should always call `super_<foo>`  in your implementation of `visit_<foo>`

Reply:
I added some docs to the vistor trait so it is clearer how it works